### PR TITLE
Add rundeck_acl_policy module

### PR DESCRIPTION
**Pull request by nerzhul**
##### SUMMARY

This module add an ansible interface to rundeck permitting to manage rundeck ACLs dynamicly through rundeck API.
Rundeck is a nice frontend tool which can trigger tasks on events (API, periodic, at like), but ACLs are very static in configuration files and rundeck doesn't reload them. API module permits to dynamicly change ACLs.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

rundeck_acl_policy

##### ANSIBLE VERSION

```
2.2.1.0
```

##### ADDITIONAL INFORMATION

```
TASK [rundeck : RUNDECK_ACL_POLICY | Manage rundeck projects] ******************
ok: [RUNDECK001] => (item={u'policy': {u'context': {u'application': u'rundeck'}, u'description': u'Admin, all access.', u'for': {u'project': [{u'allow': u'*'}], u'resource': [{u'allow': u'*'}], u'storage': [{u'allow': u'*'}], u'project_acl': [{u'allow': u'*'}]}, u'by': {u'group': u'ops'}}, u'name': u'admin_all_app_rundeck'})
ok: [RUNDECK001] => (item={u'policy': {u'context': {u'project': u'.*'}, u'description': u'Admin, all access.', u'for': {u'node': [{u'allow': u'*'}], u'job': [{u'allow': u'*'}], u'resource': [{u'allow': u'*'}], u'adhoc': [{u'allow': u'*'}]}, u'by': {u'group': u'ops'}}, u'name': u'admin_all'})
changed: [RUNDECK001] => (item={u'policy': {u'context': {u'application': u'rundeck'}, u'description': u'Company users project access', u'for': {u'project': [{u'match': {u'name': u'Plateform.*'}, u'allow': [u'read']}]}, u'by': {u'username': [u'john.doe', u'karl.johnson', u'jack.bauer'], u'group': [u'devs']}}, u'name': u'dev_project_access'})
```

### CODE REVIEW

File: lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(required=True, type='str', choices=['present', 'absent']),
+            name=dict(required=True, type='str'),
+            url=dict(required=True, type='str'),
+            api_version=dict(type='int', default=14),
+            token=dict(required=True, type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py]: Please add ` , no_log=True` so secret things are masked from user output

Comment (nerzhul) [reply lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py]: oh yeah, i missed it

**Comment by bjolivot** [comment]
Great work
shipit

**Comment by jctanner** [comment]
shipit

**Comment by nerzhul** [comment]
bot_status

File: lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+- name: Remove a rundeck system policy
+  rundeck_project:
+    name: "Project_02"
+    url: "https://rundeck.example.org"
+    token: "mytoken"
+    state: absent
+'''
+
+RETURN = '''
+changed:
```
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py]: this is a common return, please remove it.

File: lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            # If not in check mode, remove the project
+            if not self.module.check_mode:
+                self.request_rundeck_api("system/acl/%s.aclpolicy" % self.module.params["name"], method="DELETE")
+            self.module.exit_json(changed=True, before=facts, after={})
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(required=True, type='str', choices=['present', 'absent']),
```
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py]: it is common to have the default `state=present`. Consider it for consistency reasons across modules. But it is not a blocker.

**Comment by bjolivot** [comment]
shipit

File: lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            policy=dict(type='str'),
+            project=dict(type='str'),
+        ),
+        supports_check_mode=True
+    )
+
+    if module.params["api_version"] < 14:
+        module.fail_json(msg="API version should be at least 14")
+
+    if module.params["policy"] is None and module.params["state"] == "present":
```
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/rundeck_acl_policy.py]: this is a common pattern and there is some code which does the same in basic module_utils. e.g. https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/exoscale/exo_dns_record.py#L394 

FYI

**Review by resmo — CHANGES_REQUESTED** [review]

**Comment by bjolivot** [comment]
bot_status

**Comment by nerzhul** [comment]
@resmo updated !

**Comment by bjolivot** [comment]
shipit

**Comment by bjolivot** [comment]
shipit

**Review by resmo — APPROVED** [review]

**Comment by resmo** [comment]
shipit

**Comment by abadger** [comment]
I see enough shipits from people that I'm going to merge this (I know the bot is doing the right thing by resetting the shipit count each time changes are made but I'm trusting that the reviewers are just seeing the code improve.)

A few *really* minor notes that are not blockers:
* Usually we use actual comments for the license headers instead of docstrings.  docstrings would be visible as the module's help() if someone were to import the module in python and also take up memory at runtime whereas comments do not.
* We try to avoid using "_" as an identifier.  This is because we may one day want to localize the messages in our code.  gettext convention is to use "_" as the name of the function which both marks translatable strings and replaces the strings with their translated values.

Just some things to change when you next need to make changes to the module.

**Comment by abadger** [comment]
Merged to devel

**Pull request merged**